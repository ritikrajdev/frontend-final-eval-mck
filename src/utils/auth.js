export function getAuthToken() {
  return localStorage.getItem('authToken');
}

export function setAuthToken(token) {
  localStorage.setItem('authToken', token);
}

export function removeAuthToken() {
  localStorage.removeItem('authToken');
}
