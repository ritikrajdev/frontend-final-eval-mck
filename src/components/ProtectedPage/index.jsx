import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { LOGIN_ROUTE } from '../../constants/routes';
import { getAuthToken } from '../../utils/auth';

export default function ProtectedPage({ children }) {
  const navigate = useNavigate();
  const [useEffectCalled, setUseEffectCalled] = useState(false);
  useEffect(() => {
    if (!getAuthToken()) {
      navigate(LOGIN_ROUTE);
    }
    setUseEffectCalled(true);
  });
  if (!useEffectCalled) {
    return 'Initializing...';
  }
  return children;
}

ProtectedPage.propTypes = {
  children: PropTypes.node.isRequired,
};
