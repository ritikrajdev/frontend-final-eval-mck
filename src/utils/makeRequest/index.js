import axios from 'axios';
import {
  ERROR_ROUTE,
  ERROR_WITH_CODE_ROUTE,
  LOGIN_ROUTE,
} from '../../constants/routes';
import { getAuthToken, removeAuthToken } from '../auth';

export async function makeRequest(apiEndpoint, dyamicConfig, navigate) {
  dyamicConfig = dyamicConfig ?? {};

  const authToken = getAuthToken();
  if (authToken) {
    if (dyamicConfig.headers) {
      dyamicConfig.headers['Authorization'] = authToken;
    } else {
      dyamicConfig.headers = { Authorization: authToken };
    }
  }

  try {
    const response = await axios({
      ...dyamicConfig,
      ...apiEndpoint,
    });

    return response.data;
  } catch (err) {
    if (navigate) {
      if (err.response?.status) {
        if (err.response.status === 401) {
          removeAuthToken();
          navigate(LOGIN_ROUTE);
        } else {
          navigate(ERROR_WITH_CODE_ROUTE(err.response.status));
        }
      } else {
        navigate(ERROR_ROUTE);
      }
    } else {
      throw err;
    }
  }
}
