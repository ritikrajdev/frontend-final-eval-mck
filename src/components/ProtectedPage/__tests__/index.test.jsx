import { render } from '@testing-library/react';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import ProtectedPage from '..';
import { getAuthToken } from '../../../utils/auth';

jest.mock('react-router-dom');
jest.mock('../../../utils/auth');

describe('ProtectedPage', () => {
  it('should render correctly', async () => {
    getAuthToken.mockReturnValue('token');

    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);

    const { asFragment } = render(<ProtectedPage>Yo</ProtectedPage>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should redirect to login page if no auth token', async () => {
    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);

    render(<ProtectedPage>Yo</ProtectedPage>);
    expect(navigate).toBeCalledWith('/login');
  });
});
