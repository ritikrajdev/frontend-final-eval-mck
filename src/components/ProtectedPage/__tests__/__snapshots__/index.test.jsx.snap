// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedPage should render correctly 1`] = `
<DocumentFragment>
  Yo
</DocumentFragment>
`;
